analyzer/decrypt_crypto.cc

```diff
@@ -234,59 +232,53 @@ std::vector<uint8_t> decrypt(std::vector<uint8_t> client_key, std::vector<uint8_
 
 	// Setting the second parameter to NULL will pass it as Associated Data
 	EVP_CipherUpdate(ctx, NULL, &out, decryptInfo.unprotected_header.data(),
 	                 decryptInfo.unprotected_header.size());
 
 	// Set the actual data to decrypt data into the decrypt_buffer. The amount of
 	// byte decrypted is stored into `out`
-	EVP_CipherUpdate(ctx, decrypt_buffer, &out, encrypted_payload.data(), encrypted_payload.size());
+	EVP_CipherUpdate(ctx, decrypt_buffer.data(), &out, encrypted_payload, encrypted_payload_size);
 
 	// Validate whether the decryption was successful or not
 	EVP_CipherFinal_ex(ctx, NULL, &out2);
 	EVP_CIPHER_CTX_free(ctx);
 
-	// Copy the decrypted data from the decrypted buffer into a new vector and return this
-	// Use the `out` variable to only include relevant bytes
-	std::vector<uint8_t> decrypted_data(decrypt_buffer, decrypt_buffer + out);
-	return decrypted_data;
+	// Copy the decrypted data from the decrypted buffer into a Bytes instance.
+	return hilti::rt::Bytes(decrypt_buffer.data(), decrypt_buffer.data() + out);
+	}
+
 	}
 
 /*
 Function that is called from Spicy. It's a wrapper around `process_data`;
 it stores all the passed data in a global struct and then calls `process_data`,
 which will eventually return the decrypted data and pass it back to Spicy.
 */
-hilti::rt::Bytes decrypt_crypto_payload(const hilti::rt::Bytes& entire_packet,
-                                        const hilti::rt::Bytes& connection_id,
-                                        const hilti::rt::integer::safe<uint64_t>& encrypted_offset,
-                                        const hilti::rt::integer::safe<uint64_t>& payload_length,
-                                        const hilti::rt::Bool& from_client)
+hilti::rt::Bytes
+QUIC_decrypt_crypto_payload(const hilti::rt::stream::SafeConstIterator& packet_stream,
+                            const hilti::rt::Bytes& connection_id,
+                            const hilti::rt::integer::safe<uint64_t>& encrypted_offset,
+                            const hilti::rt::integer::safe<uint64_t>& payload_length,
+                            const hilti::rt::Bool& from_client)
 	{
 
 	if ( payload_length < 20 )
 		throw hilti::rt::RuntimeError(hilti::rt::fmt("payload too small %ld < 20", payload_length));
 
-	if ( entire_packet.size() < encrypted_offset + payload_length )
-		throw hilti::rt::RuntimeError(hilti::rt::fmt(
-			"packet too small %ld < %ld", entire_packet.size(), encrypted_offset + payload_length));
+	if ( (packet_stream + (encrypted_offset + payload_length - 1)).isEnd() )
+		throw hilti::rt::RuntimeError(
+			hilti::rt::fmt("packet too small %ld", encrypted_offset + payload_length));
 
 	// Fill in the entire packet bytes
 	std::vector<uint8_t> e_pkt;
-	for ( const auto& singlebyte : entire_packet )
-		{
-		e_pkt.push_back(singlebyte);
-		}
+	hilti::rt::stream::detail::UnsafeConstIterator it{packet_stream};
```

Comment:
Related to my previous comment, you again rely on the fact that `packet_stream` has enough data, else something will break catastrophically, all so you can pass a less constrained type from Spicy.